controllers/reconciler.go

```diff
@@ -0,0 +1,336 @@
+	if equality.Semantic.DeepEqual(orig, obj) {
+		return nil
+	}
+
+	// Order matters.
+	// First need to update the object, and then update the status.
+	// So dependent resources won't see READY before it has been updated with new values
+
+	// Now we can update the status
+	return retry.RetryOnConflict(retry.DefaultRetry, func() error {
+		// When updated, object status is vanished.
+		// So we waste a copy for that,
+		// while the original object must already have all the fields updated in runtime
+		// Additionally, it gets the "latest version" to resolve optimistic concurrency control conflict
+		latest := obj.DeepCopyObject().(client.Object)
+		if err := r.Get(ctx, types.NamespacedName{Name: latest.GetName(), Namespace: latest.GetNamespace()}, latest); err != nil {
+			return err
+		}
+
+		updated := obj.DeepCopyObject().(client.Object)
+		updated.SetResourceVersion(latest.GetResourceVersion())
+		err := r.Update(ctx, updated)
+		if err != nil {
+			return err
+		}
+
+		obj.SetResourceVersion(updated.GetResourceVersion())
+		return r.Status().Update(ctx, obj)
+	})
+}
+
+func (r *Reconciler[T]) resolveToken(ctx context.Context, obj T) (string, error) {
+	if r.DefaultToken != "" {
+		return r.DefaultToken, nil
+	}
+
+	auth := obj.AuthSecretRef()
+	if auth == nil {
+		return "", errNoTokenProvided
+	}
+
+	clientAuthSecret := &corev1.Secret{}
+	if err := r.Get(ctx, types.NamespacedName{Name: auth.Name, Namespace: obj.GetNamespace()}, clientAuthSecret); err != nil {
+		r.Recorder.Eventf(obj, corev1.EventTypeWarning, eventUnableToGetAuthSecret, err.Error())
+		return "", fmt.Errorf("cannot get secret %q: %w", auth.Name, err)
+	}
+
+	return string(clientAuthSecret.Data[auth.Key]), nil
+}
+
+func (r *Reconciler[T]) newAivenClient(ctx context.Context, obj T) (avngen.Client, error) {
+	token, err := r.resolveToken(ctx, obj)
+	if err != nil {
+		return nil, err
+	}
+
+	avnGen, err := NewAivenGeneratedClient(token, r.KubeVersion, r.OperatorVersion)
+	if err != nil {
+		r.Recorder.Event(obj, corev1.EventTypeWarning, eventUnableToCreateClient, err.Error())
+		return nil, fmt.Errorf("cannot initialize aiven generated client: %w", err)
+	}
+
+	return avnGen, nil
+}
+
+func (r *Reconciler[T]) createResource(ctx context.Context, controller AivenController[T], obj T) (ctrl.Result, error) {
+	r.Recorder.Event(obj, corev1.EventTypeNormal, eventCreateOrUpdatedAtAiven, "about to create instance at aiven")
+
+	if err := controller.Create(ctx, obj); err != nil {
+		r.Recorder.Event(obj, corev1.EventTypeWarning, eventUnableToCreateOrUpdateAtAiven, err.Error())
+		meta.SetStatusCondition(obj.Conditions(), getErrorCondition(errConditionCreateOrUpdate, err))
+		return ctrl.Result{}, fmt.Errorf("unable to create or update instance at aiven: %w", err)
+	}
+
+	r.Recorder.Event(obj, corev1.EventTypeNormal, eventCreatedOrUpdatedAtAiven, "instance was created at aiven but may not be running yet")
+	return ctrl.Result{RequeueAfter: pollInterval}, nil
```

Comment:
We should save the secret here as well.

Reply:
After `Create` we now also pass `CreateResult.SecretDetails` into `publishSecretDetails`, so connection secrets are written both on create and on update, not just in the update path.